rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,370 @@
+file pruning with file-level column stats. For the example above, even in the new partitioning scheme, without data
+being physically partitioned by the datestr, the data can still be co-located based the date, because of natural
+ingestion order or Hudi's clustering operation. In this case, we can effectively prune files based on the range of
+datestr for each file, regardless how the files are stored under different physical partition paths in the table. We
+need to add such file pruning support in Hudi.
+
+## Goals and Non-Goals
+
+Based on the use cases we plan to support, we set the following goals and non-goals for this RFC:
+
+### Goals
+
+- Support file pruning on functions defined on data column
+- Improve the co-location of data without folder structure for efficient file pruning
+- Improve the query performance based on the new pruning capability compared existing approach in Hudi
+
+### Non-Goals
+
+- DO NOT remove physical partitioning, which remains as an option for physically storing data in different folders and
+  partition pruning
+- DO NOT tackle the support of partition evolution in this RFC, but make sure the file pruning on functions defined on
+  data column can be applied on the partition evolution use case
+
+## Design and Implementation
+
+To achieve the goals above, in this RFC, we introduce the **Index Function**, a function defined on data column which is
+indexed to facilitate file pruning, which works at high level as follows:
+
+1. User specifies the Index Function, including the original data column, the expression applying a function on the
+   column, and optionally the new column name, through SQL or Hudi's write config. One or more Index Functions can be
+   added or dropped for both new and existing Hudi tables at any time.
+2. The Index Function is registered to the table in the table properties, to keep track of the relationship between the
+   data column and the Index Function.
+3. Hudi creates the index metadata, such as file-level column stats, of Index Functions in metadata table. Hudi marks
+   whether one Index Function and the index metadata are ready for use in file pruning.
+4. No data is written to the data files for the Index Function, i.e., the data files do not contain a new data column
+   corresponding to the Index Function, to save storage space.
+5. When query engine makes a logical plan for a query, Hudi intercepts the predicate that relates to an Index Function,
+   looks for the corresponding index in metadata table, and prunes files based on the index.
+
+When writing data to a non-partitioned table or data files under one partition path, selective Index Functions can be
+specified as ordering field during write operations. For inserts, Hudi's write client sorts or buckets the data based on
+the Index Function(s) so that each data file has a narrow and mostly non-overlapping range of values for the
```

Comment:
How to keep data in order over time. rely on full table sortby  clustering?

Reply:
yes probably.. unless we sort on fly (HoodieSortedMergeHandle). Will file a followup for this.